x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RemoteAccessAuthenticationService.java

```diff
@@ -0,0 +1,161 @@
+
+import static org.elasticsearch.core.Strings.format;
+import static org.elasticsearch.xpack.core.security.authc.RemoteAccessAuthentication.REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY;
+import static org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor.REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY;
+
+public class RemoteAccessAuthenticationService {
+
+    private static final RoleDescriptor CROSS_CLUSTER_SEARCH_ROLE = new RoleDescriptor(
+        "_cross_cluster_search",
+        new String[] { ClusterStateAction.NAME },
+        null,
+        null,
+        null,
+        null,
+        null,
+        null,
+        null
+    );
+    private static final Logger logger = LogManager.getLogger(RemoteAccessAuthenticationService.class);
+
+    private final AuthenticationService authenticationService;
+
+    public RemoteAccessAuthenticationService(AuthenticationService authenticationService) {
+        this.authenticationService = authenticationService;
+    }
+
+    public void authenticate(
+        final String action,
+        final TransportRequest request,
+        final boolean allowAnonymous,
+        final ActionListener<Authentication> listener
+    ) {
+        final Authenticator.Context authcContext = authenticationService.newContext(action, request, allowAnonymous);
+        final ThreadContext threadContext = authcContext.getThreadContext();
+
+        if (threadContext.getHeader(AuthenticationField.AUTHENTICATION_KEY) != null) {
+            withRequestProcessingFailure(
+                authcContext,
+                new IllegalArgumentException("authentication header is not allowed with remote access"),
+                listener
+            );
+            return;
+        }
+
+        final RemoteAccessHeaders remoteAccessHeaders;
+        try {
+            remoteAccessHeaders = RemoteAccessHeaders.readFromContext(threadContext);
+        } catch (Exception ex) {
+            withRequestProcessingFailure(authcContext, ex, listener);
+            return;
+        }
+
+        try (
+            ThreadContext.StoredContext ignored = threadContext.newStoredContext(
+                Collections.emptyList(),
+                // drop remote access authentication headers since we've read their values, and we want to maintain the invariant that
+                // either the remote access authentication header is in the context, or the authentication header, but not both
+                List.of(REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY, REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY)
+            )
+        ) {
+            final Supplier<ThreadContext.StoredContext> storedContextSupplier = threadContext.newRestorableContext(true);
+            // Write remote access credential to the Authorization header, so we can re-use generic authentication service functionality for
+            // authc token extraction
+            threadContext.putHeader("Authorization", remoteAccessHeaders.clusterCredentialHeader());
```

Comment:
With this approach, how do we enforce the cluster credential to be API keys only (for the timebeing at least)?

Reply:
Ah whoops! We have an assertion inside `toRemoteAccess` that the base authc has to be via API key but of course that doesn't prevent this behavior outside of tests. 

I'm considering two possibilities here:

**Option 1:** insert a validation after authentication runs, s.th. like:

```
authenticationService.authenticate(
  authcContext,
  new ContextPreservingActionListener<>(storedContextSupplier, ActionListener.wrap(authentication -> {
    if (false == authentication.isApiKey()) {
       listener.onFailure(...)
       return;
    }  
    final RemoteAccessAuthentication remoteAccessAuthentication = remoteAccessHeaders.remoteAccessAuthentication();
      
  ...
}
```

This is generic, i.e., it works without inspecting the header through an API-key specific lens. The downside is that we don't fail as early as possible since the first authentication step still executes. This runs counter to our security goal of locking down cross cluster access as much as possible. A malicious querying cluster could still attempt to force similar behavior (i.e., trying to authenticate through arbitrary credentials) by going through the HTTP interface but I don't think there is complete parity here. This might be far-fetched, but there could be configurations where the HTTP interface is locked down differently (e.g., only accessible from a VPN) than the remote access interface.

**Option 2:** specialize authc token extraction inside RemoteAccessAuthcService, and don't use Authz header

The idea would be to "leak" API-key specific details to `RemoteAccessAuthcService`, s.th. like:

```
AuthcToken apiKeyToken = apiKeyService.getCredentialsFromHeader(remoteAccessHeaders.clusterCredentialHeader()) // this is a new method, and RemoteAccessAuthcService has direct access to ApiKeyService
if (apiKeyToken == null) {
  // fail request
}
authcContext.addAuthenticationToken(apiKeyToken);
authenticationService.authenticate(...)
```

This is counter to my initial preference, but I'm leaning towards it now since it allows us to fail early and prevents a malicious querying cluster from attempting arbitrary authentication via the remote access port.

The above could further be refactored to be more generic but I don't think this is warranted since there are no current plans to support authc methods other than API keys.

Reply:
I also prefer opiton 2 which was more or less the previous idea. I personally think token extraction and authentication are two separate tasks. We do not need to delegate both to the AuthenticationService. This is similar to how SAML (or OIDC) realm authentication works, e.g. the transport action extract the token and pass it to AuthenticationService for authentication.

With this approach, we may want to add another field to the `Context` to prevent `run-as`. Alternatively this can be prevented by the caller. I am ok with either.

Reply:
> With this approach, we may want to add another field to the Context to prevent run-as. Alternatively this can be prevented by the caller. I am ok with either.

Good call! I'll add this (most likely to the caller).

Reply:
I've added a validation that there is no `run-as` header -- is this sufficient or am I missing part of your suggestion here?